main/resource-access.bs

```diff
@@ -70,7 +70,99 @@ A Solid data pod MUST conform to the Web Access Control specification [[!WAC]].
 
-Issue: Write Cross-Origin Resource Sharing section.
-
-Draft:
-A Solid data pod MUST conform to the CORS specification [[!FETCH]].
+### Background and Need ### {#cors-need}
+<em>This section is non-normative.</em>
+
+[=Solid apps=] typically access data from multiple sources.
+However,
+Web browsers by default prevent apps that run on one origin
+from accessing data on other origins.
+This cross-origin protection is a security mechanism
+that ensures malicious websites cannot simply read
+your profile or banking details from other websites.
+However, this reasonable default poses a problem
+even for benevolent Solid apps,
+which might have good reasons to access data from different places.
+For instance,
+a Solid app at `https://app.example/`
+would be prevented from accessing data on
+`https://alice-data-pod.example/` or `https://bob-data-pod.example/`,
+even when Alice and Bob have given the user of the app
+their permission to see some of their data.
+
+For cases where the other origins
+have their own access protection mechanism—
+[like within Solid](#wac)—
+the browser's built-in cross-origin protection
+is actually an obstacle rather than a feature.
+After all,
+[=data pods=] already ensure through access control
+that certain documents can only be accessed by specific people.
+Preventively blocking apps from different origins
+thus introduces an unnecessary barrier.
+
+Fortunately,
+Web servers can indicate to the browser
+that certain documents do not require cross-origin protection.
+This mechanism to selectively disable that protection
+is called *Cross-Origin Resource Sharing* or *CORS* [[FETCH]].
+By responding to browser requests
+with a specific combination of HTTP headers,
+servers can indicate which actions are allowed for a given resource.
+For a Solid data pod,
+the goal is to allow *all* actions on the CORS level,
+such that the deeper [access control layer](#wac)
+can exert full control over the app's allowed permissions.
+The next section describes how to achieve this
+through the right HTTP header configuration.
+
+
+### Required server-side implementation ### {#cors-server}
+
+A [=data pod=] MUST implement the CORS protocol [[!FETCH]]
+such that, to the extent possible,
+the browser does not block
+any request sent to it or response received from it.
+If the data pod wishes to block access to a resource,
+this MUST NOT happen via CORS
+but MUST instead be communicated to the Solid app in the browser
+through HTTP status codes such as
+`401`, `403`, or `404` [[!RFC7231]].
+Furthermore,
+all response headers MUST be exposed to the Solid app.
+
+Note: Since the CORS protocol is part of a Living Standard,
+it should still be considered in evolution.
+Changes to the CORS protocol might thus happen at any point,
+which might necessitate changes to server implementations
+for continued prevention of undesired blocking.
+A [proposal](https://github.com/whatwg/fetch/issues/878) to mitigate this
+has been suggested.
+
```

Comment:
This Note is perhaps not significant enough for this document.

Reply:
It depends. We don't have to link to the issue as far as I'm concerned, but it seems crucial to me that people know that there is no such thing as opting out of same-origin protections, that servers will periodically need to be updated because CORS is a living spec.

The link to the issue was to indicate that, perhaps in the future, such periodical updates might not be necessary anymore.

I feel strongly about keeping in the first part, less strongly about the second part. What do you think?

Reply:
Good. Let's keep both for now.

Reply:
The information in this Note is useful for Security Considerations. What do you think about moving it there?

Reply:
It's not a security issue, rather the reverse. CORS will arbitrarily add _more_ security in the future, blocking functionality but not security.

Reply:
The point was to say something along the lines of "CORS is a LS, so take care with anything pertaining to security" - we can't warn against unknown items at this time, so it is a general warning.

Reply:
> CORS is a LS, so take care with anything pertaining to security

So the "CORS is a LS" is in there.
"take care with anything pertaining to security" is, purposely, not: CORS is only going to get stricter with regard to security, not loser. So there are no security problems, and hence I wouldn't move this to the security section either. There are only "reverse" security problems: when the standard makes things more strict so you can't see things you should be able to.

Reply:
It is risky for us to assume anything about a mutating spec's security implications.

A general warning about potential security implications of LSs would suffice.

Reply:
Good idea, added through 3912ad14aefac4a761160dcd74f612c40c97a1fd.